Source/ASDisplayTree.h

```diff
@@ -0,0 +1,112 @@
+//
+//  ASDisplayTree.h
+//  AsyncDisplayKit
+//
+//  Created by Adlai on 5/28/18.
+//  Copyright © 2018 Pinterest. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <AsyncDisplayKit/ASBaseDefines.h>
+#import <AsyncDisplayKit/ASTraitCollection.h>
+
+@class ASDisplayNode;
+
+typedef NS_OPTIONS(NSUInteger, ASTreeEnumerationOptions) {
+  ASEnumerateBreadthFirst  = 1 << 0,  // Default is depth-first.
+  ASEnumerateSkipSelf   = 1 << 1,     // Default is to include self.
+};
+
+typedef NS_ENUM(NSInteger, ASTreeInsertionLocation) {
+  ASTreeInsertBelow,      // Other node is sibling above. Index ignored.
+  ASTreeInsertReplace,    // Other node is replaced. Index ignored.
+  ASTreeInsertAbove,      // Other node is sibling below. Index ignored.
+  ASTreeInsertAtEnd,      // Other node is parent. Index ignored.
+  ASTreeInsertAtIndex     // Other node is parent.
+};
+
+NS_ASSUME_NONNULL_BEGIN
+
+/*
+ * An entire tree of display nodes. Subtrees are not
+ * represented by these objects.
+ *
+ * When a node is added to a tree, the tree object
+ * is pushed down to it from its new parent.
+ *
+ * Trees use a recursive lock to ensure safe access
+ * from multiple threads simultaneously.
+ *
+ * Trees do not own their nodes. The root node is responsible
+ * for owning the tree.
+ */
+AS_SUBCLASSING_RESTRICTED
+@interface ASDisplayTree : NSObject <NSLocking>
+
+#pragma mark - Querying
+
+/*
+ * Copy the direct subnodes of the given node.
+ * Consider using @p enumerateSubnodesOf: @c instead.
+ */
+- (NSArray<ASDisplayNode *> *)copySubnodesOf:(ASDisplayNode *)node;
```

Comment:
I wonder if we should create a protocol like <ASDisplayHierarchy> or something that specifies the ability to walk the tree, copy subnodes, etc. Not necessary, just an idea to consider. Could have some relationship to <ASLayoutElement>.